src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs

```diff
@@ -180,8 +180,10 @@ private BoundExpression VisitExpressionImpl(BoundExpression node)
             // like compound assignment does (extra flag only passed when it is an expression
             // statement means that this constraint is not violated).
             // Dynamic type will be erased in emit phase. It is considered equivalent to Object in lowered bound trees.
+            // stackalloc bound nodes have no type, and they are updated to the appropriate type after lowering
             Debug.Assert(visited == null || visited.HasErrors || ReferenceEquals(visited.Type, node.Type) ||
```

Comment:
I think you can preserve the assertion unchanged if you lower the conversion node rather than the stackalloc node. That would also allow you to remove the (now no-op) conversion.